### DISCUSSION TITLE
[Feature Idea/Request] Add Amazon Bedrock as a provider in Kong AI Proxy

### DISCUSSION THREAD

**Discussion by kandakji** [Kong/kong#12631] [Ideas and feature requests]
Hi,

I'm a Data & AI Solutions Architect at AWS. Enterprise customers are adopting LLM/FM gateways to facilitate project Time-to-market and to unify and standardize monitoring, operations, security, and app integration.

Amazon Bedrock is one of the most popular services on AWS that enterprise customers, like Intuit, Adidas, Booking.com, Nasdaq and Salesforce, use to access models from 7 providers (**AI21 Labs**, **Anthropic**, **Cohere**, **Meta**, **Mistral AI**, **Stability AI**, and **Amazon**) as a completely serverless fully-managed service.

We (fellow solutions architects and myself) would be happy to contribute to integrating Amazon Bedrock within Kong's AI Proxy to facilitate adoption of this plugin for customers and users using Amazon Bedrock on AWS (in case Bedrock is not within your roadmap already).

**Comment by hanshuebner** [comment]
@tysoekong Is Amazon Bedrock something that is on our roadmap already?

**Comment by subnetmarco** [comment]
Yes, we would like to have AWS Bedrock. If you would like to contribute, we can offer guidance. Is it still something that you would still like to do? @kandakji

**Reply by kandakji** [reply]
Hi Marco, yes, I'd be happy to start working on the contribution with some colleagues by the beginning of next week. What kind of guidance can you provide us ?

**Comment by tysoekong** [comment]
Hey @kandakji - I think we are good to get working on this! Any help is always amazing.

We can meet ASAP and I can show you all how the interface works? It should be simple as:

1. A new file to handle chosen Bedrock inference formats (just a "sub-class" type Lua file that implements the required functions)
2. Plug that interface into the config options so that users to activate it

**Reply by tysoekong** [reply]
I predict only... 1 hour call, to go into it, alternatively I can just email you the interface docs itself and read in your own time?

**Reply by Tieske** [reply]
me too 😄

**Comment by henriquelucasdf** [comment]
Hey folks, how's it going?

Firstly, I'd like to congratulate and thank the entire team involved with the AI Proxy plugin, it's a fantastic feature.

I'd like to ask if there's any news regarding integration with Amazon Bedrock? Is there any place where I can follow the development or try to contribute?

**Reply by subnetmarco** [reply]
It's currently prioritized and in the works, will ship in 3.8 in August but most likely you will be able to compile a beta by end of June.

**Comment by subnetmarco** [comment]
The AWS Bedrock PR has been opened, needs to be reviewed and then it will be merged: https://github.com/Kong/kong/pull/13054

We also added support for Google Gemini by the way, which also needs to be reviewed before being merged: https://github.com/Kong/kong/pull/12948

**Comment by mehmetaltuntas** [comment]
As far as I am aware, AWS Bedrock authorization flow requires "Signature v4" using the secret, access keys. Does Kong API Gateway look after this? If so, does it use any library/SDK somehow? Thanks.

**Reply by tysoekong** [reply]
Hey there, Kong gateway manages the authentication to AWS/Bedrock. It supports the normal auth chain as the AWS SDKs, so you can assign an IAM Role to the pod/deployment, EC2 instances, Fargate execution role, directly to the Kong process.

You can also hard-code IAM User access+secret keys into either the environment variables of the Kong process, or set them in the plugin configuration.

From the client, you **currently** (Kong 3.9 and below) can use any OpenAI-compatible '/chat/completions' library/framework. This includes multi-modal.

That can be OpenAI SDK, Langchain, or anything else with OpenAI-format adapter.

A future revision of Kong will add the support to use any boto3 SDK with "invoke-model" or "converse" operations, however the authentication still has to be controlled at the Kong gateway (we cannot pass client-signed requests directly to Bedrock without breaking the signature).

**Reply by mehmetaltuntas** [reply]
Thanks for the reply. I did implement both ( Open AI n Bedrock ) using different API Gateway service provider and the challenge was to have the Bedrock credentials to be signed as expected.  I ended up having a simple API app that does the signature as well as the OpenAI-format conversion so that the output would remain consistent.  I am assuming this process is now handled by Kong, right?

**Comment by subnetmarco** [comment]
Closing this discussion since AWS Bedrock has been a supported LLM provider for a while.
